### PR TITLE
feat: Add logging service

### LINKED ISSUE #36: feat: Add log service
# References

- https://docs.nestjs.com/techniques/logger

### PR DESCRIPTION
**Pull request by kimseungbin**
# Description

- NestJS built-in logger 를 사용해 HTTP req/res 를 기록
- Console 로 로그를 기록하여 추후 ECS 상의 로그를 CloudWatch 로 전달할 예정

- Close #36

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,6 +16,18 @@ docker compose up
 
 *TBD*
 
+# Logging
+
+Logging on HTTP `request` and `response` is implemented as a *Middleware function* and followed the official course.
+
+The format follows: `${method} ${originalUrl} ${statusCode} ${contentLength} - ${userAgent} ${ip} ${traceId}`
```
Comment (eunkyui) [inline README.md:23]: 에러가 발생할 때는 좀 더 자세히 로그를 남겨주는 방식도 고민해보면 좋을 것 같네요

File: README.md [hunk 2]
```diff
@@ -16,6 +16,18 @@ docker compose up
 
 *TBD*
 
+# Logging
+
+Logging on HTTP `request` and `response` is implemented as a *Middleware function* and followed the official course.
```
Comment (eunkyui) [inline README.md:21]: Middleware 를 사용하는 방식과 Interceptor 등 NestJS 의 다른 기능을 사용하는 방식의 차이가 어떻게 되는지도 한번 찾아보면 좋을 것 같네요

File: README.md (hunk 1 above)
Comment (kimseungbin) [reply README.md:23]: 1. log level 이 error 일 경우 에러 객체 전달
2. Format 에 대한 고민
3. res 에도 error 에 대한 정보 전달

- 환경변수를 이용해서 log 를 JSON 으로도 출력 가능하게 변경하고 문서화

File: README.md (hunk 2 above)
Comment (kimseungbin) [reply README.md:21]: In NestJS, both Interceptors and Middleware can be used to handle logging, but the way each of them is used, and their capabilities are slightly different:
Interceptors are a part of Aspect-Oriented Programming (AOP) principles in NestJS. They have a well-defined place in the request/response cycle where they run - namely, right before a method is executed (before), and after its result (after). Interceptors are perfectly suited to add behavior to methods, such as transforming the result or add additional logic like logging. Interceptors also have access to the execution context, which gives them the advantage of understanding which controller and which method is being executed, and the ability to modify the result of the execution. Lastly, they are coupled to the NestJS ecosystem and might only be used locally (single route) or globally.
Middleware in NestJS act in a similar way as they would in express. They are functions that have access to the request object (req), the response object (res), and the next() function in the application’s request-response cycle. Middleware functions can perform the following tasks:
Execute any code.
Make changes to the request and the response objects.
End the request-response cycle.
Call the next middleware in the stack. Middleware can therefore be used for logging, and they work on the request level. However, they do not understand which controller would handle a request, let alone which method. Middleware can also be applied to every route in your application or only to specified routes.
In the context of logging, therefore:
If you are interested in logging the raw details of requests and responses (like request headers, response status, etc.), then Middleware might be the more straightforward option.
If you want your logging to be more granular, and to know which controller and which method is handling a particular request, then Interceptors would be more suitable as they have access to the execution context.
Each case is a valid way to implement logging in a NestJS application and your choice will depend on your specific needs and requirements.

JetBrains AI

Comment (kimseungbin) [reply README.md:21]: https://docs.nestjs.com/interceptors#aspect-interception

**Pull request merged**